libs/gl-client/src/signer/mod.rs

```diff
@@ -761,16 +766,145 @@ impl Signer {
     ) -> Result<(), anyhow::Error> {
         let scheduler = self.init_scheduler(scheduler_uri).await?;
         tokio::select! {
-            run_forever_inner_res = self.run_forever_inner(scheduler) => {
+            run_forever_inner_res = self.run_forever_inner(scheduler.clone()) => {
                 error!("Inner signer loop exited unexpectedly: {run_forever_inner_res:?}");
             },
+            run_forever_scheduler_res = self.run_forever_scheduler(scheduler) => {
+                error!("Scheduler signer loop exited unexpectedly: {run_forever_scheduler_res:?}")
+            }
             _ = shutdown.recv() => debug!("Received the signal to exit the signer loop")
         };
 
         info!("Exiting the signer loop");
         Ok(())
     }
 
+    async fn run_forever_scheduler(
+        &self,
+        mut scheduler: SchedulerClient<tonic::transport::Channel>,
+    ) -> Result<(), anyhow::Error> {
+        let (sender, rx) = mpsc::channel(1);
+        let outbound = ReceiverStream::new(rx);
+        // let inbound_future = scheduler.signer_requests_stream(outbound);
+
+        let mut stream = scheduler
+            .signer_requests_stream(outbound)
+            .await?
+            .into_inner();
+
+        trace!("Starting to stream signer requests from scheduler");
+
+        loop {
+            match stream.message().await {
+                Ok(Some(msg)) => {
+                    let req_id = msg.request_id;
+                    trace!("Processing scheduler request {}", req_id);
+                    match msg.request {
+                        Some(signer_request::Request::ApprovePairing(req)) => {
+                            self.process_pairing_approval(req_id, req, sender.clone())
+                                .await
+                        }
+                        None => {
+                            debug!("Received an empty signing request");
+                        }
+                    };
+                }
+                Ok(None) => {
+                    debug!("End of stream, this should not happen by the server");
+                    return Err(anyhow!("Scheduler closed the stream"));
+                }
+                Err(e) => {
+                    debug!("Got an error from the scheduler {}", e);
+                    return Err(anyhow!("Scheduler stream error {e:?}"));
+                }
+            };
+        }
+    }
+
+    async fn process_pairing_approval(
+        &self,
+        req_id: u32,
+        req: ApprovePairingRequest,
+        stream: mpsc::Sender<SignerResponse>,
+    ) -> () {
```

Comment:
The empty return makes the following construction necessary:

```
debug!(...);
return;
```

Which could have been an inner function returning a real `Result<>`, thus only having to do the logging once, at a level decided once, and we could skip the `return`. Might be a bit more rustic.